Geometry: add a Geometry reader for various supported mesh files

tamiko opened the pull request:
Extracted from #130

gregorychristian committed:
Geometry: add a geometry that reads in GMSH *.msh files

tamiko committed:
Add Gregory to the list of contributors

tamiko commented on source/geometry_gmsh.h:
```diff
@@ -0,0 +1,42 @@
+          : Geometry<dim>("gmsh", subsection)
+      {
+        filename_ = "ryujin.msh";
+        this->add_parameter("filename", filename_, "GMSH *.msh file to load");
```
@gregorychristian I have changed your commit slightly by adding a parameter for the file that should be read in and renamed the class to `GMSH`.

gregorychristian replied:
@tamiko Great thanks.

ejtovar commented:
@tamiko @gregorychristian Is it possible to generalize this GMSH reader to include UCD meshes as well? That is, depending on the type of mesh that gets inputted, the reader decides if it is GMSH vs UCD? See: https://www.dealii.org/current/doxygen/deal.II/classGridIn.html#abbb16aa0aa53bfa9231b69278d397616

tamiko commented on source/geometry_gmsh.h:
```diff
@@ -0,0 +1,42 @@
+        dealii::GridIn<dim> gridin;
+        gridin.attach_triangulation(triangulation);
+        std::ifstream file(filename_);
+        gridin.read_msh(file);
```
The logic is so small that we can simply do the call to `GridIn` in the `create_triangulation()` function.

(The split that you copied where the implementation is split out into a `GridGenerator` function served a special purpose to be closer to the deal.II `GridGenerator` functions. But that isn't needed here.)

tamiko commented on README.md:
```diff
@@ -132,4 +132,5 @@ Contributors
  - Seth Gerberding ([@Gerbeset](https://github.com/Gerbeset)), Texas A&M University, TX, USA
  - David Pecoraro ([@ChrisPec27](https://github.com/ChrisPec27)), Texas A&M University, TX, USA
  - Madison Sheridan ([@Helblindi](https://github.com/Helblindi)), Texas A&M University, TX, USA
+ - Gregory Christian ([@gregorychristian](https://github.com/gregorychristian), Imperial College London, UK
```
@gregorychristian If you don't mind we'd like to add you as contributor to the list.

gregorychristian replied:
@tamiko Yes of course!

tamiko reviewed (commented):
@gregorychristian I have split out the GMSH part into a separate pull request.

tamiko commented:
@ejtovar Excellent point. I went ahead and refactored the class to be called `Reader` and switched to `GreadIn::read()` that supports various different mesh formats.
